### PR TITLE
Ring buffer completion passes incorrect length

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Resolves: #2656 

## Description

1. Ringbuffer async completion passes the wrong length for the output buffer.
2. IO Manager mock doesn't correctly copy the buffers to and from kernel mode in the async path.
3. Add instrumentation to test to catch duplicate records being returned.

## Testing

Fixed existing tests to correctly catch this.

## Documentation

No.

### CODE REVIEW

File: tests/end_to_end/test_helper.cpp [hunk 1]
```diff
@@ -281,6 +290,20 @@ _complete_overlapped(_Inout_ void* context, size_t output_buffer_length, ebpf_re
 {
     UNREFERENCED_PARAMETER(output_buffer_length);
     auto overlapped = reinterpret_cast<OVERLAPPED*>(context);
+
+    // Copy the output buffer to the user buffer.
+    if (overlapped) {
+        std::unique_lock lock(_overlapped_buffers_mutex);
+        auto it = _overlapped_buffers.find(overlapped);
+        if (it == _overlapped_buffers.end()) {
+            throw std::runtime_error("Overlapped buffer not found");
```
Comment (dthaler) [inline tests/end_to_end/test_helper.cpp:299]: why throw rather than REQUIRE()?

File: tests/end_to_end/test_helper.cpp [hunk 2]
```diff
@@ -439,10 +462,26 @@ GlueDeviceIoControl(
...
     void* local_output_buffer = nullptr;
 
+    // To correctly emulate the kernel execution context, we need to use the same buffer
+    // for both input and output.  So we allocate a buffer that is large enough to hold
+    // either the input or output, and then use that buffer for both.
+    if (overlapped) {
+        std::unique_lock lock(_overlapped_buffers_mutex);
+        auto it = _overlapped_buffers.find(overlapped);
+        if (it != _overlapped_buffers.end()) {
+            throw std::runtime_error("Overlapped buffer already in use");
```
Comment (dthaler) [inline tests/end_to_end/test_helper.cpp:475]: why throw rather than REQUIRE()?

File: tests/end_to_end/test_helper.cpp [hunk 3]
```diff
@@ -718,6 +757,15 @@ _test_helper_end_to_end::~_test_helper_end_to_end()
         _unload_all_native_modules();
 
         clear_program_info_cache();
+
+        {
+            std::unique_lock lock(_overlapped_buffers_mutex);
+            if (!_overlapped_buffers.empty()) {
+                std::cerr << "Overlapped buffers not empty" << std::endl;
```
Comment (dthaler) [inline tests/end_to_end/test_helper.cpp:764]: Is this just for informational purposes?  (i.e., not an error?)  Maybe add a // comment

File: tests/libs/common/common_tests.cpp [hunk 4]
```diff
@@ -155,7 +155,15 @@ ring_buffer_test_event_handler(_Inout_ void* ctx, _In_opt_ const void* data, siz
...
     auto it = std::find(records->begin(), records->end(), event_record);
     if (it != records->end()) {
-        event_context->matched_entry_count++;
+        auto index = std::distance(records->begin(), it);
+        if (event_context->event_received.find(index) != event_context->event_received.end()) {
+            // Duplicate event record received.
+            // Return error so that no further callback is made.
+            return -1;
+        } else {
+            event_context->event_received.insert(index);
```
Comment (shankarseal) [inline tests/libs/common/common_tests.cpp:164]: alternatively, remove the matched entry from records ?

Comment (Alan-Jowett) [reply tests/libs/common/common_tests.cpp:164]: Awesome suggestion! That makes it a lot simpler.

Comment (Alan-Jowett) [reply tests/libs/common/common_tests.cpp:164]: Actually, that doesn't work. The function bindmonitor_ring_buffer_test assumes the list of records is immutable.

File: tests/end_to_end/test_helper.cpp (hunk 3 above)
Comment (Alan-Jowett) [reply tests/end_to_end/test_helper.cpp]: This was part of debugging early on. Removed.

File: tests/socket/socket_tests.cpp [hunk 5]
```diff
@@ -269,7 +269,10 @@ connection_monitor_test(
 
     // Ring buffer event callback context.
     std::unique_ptr<ring_buffer_test_event_context_t> context = std::make_unique<ring_buffer_test_event_context_t>();
-    context->test_event_count = disconnect ? 4 : 2;
```
Comment (shankarseal) [inline tests/socket/socket_tests.cpp:272]: what about the ports on the expected records for connection_established event?

**Review by dthaler — APPROVED** [review]

File: tests/socket/socket_tests.cpp (hunk 5 above)
Comment (Alan-Jowett) [reply tests/socket/socket_tests.cpp:272]: Tracked by #2706

**Review by dv-msft — APPROVED** [review]
